File: apps/flutter-app/lib/features/tasks/data/models/task_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../../../shared/widgets/atoms/app_status_badge.dart';
import '../../domain/entities/task.dart';

/// Modèle de données Firestore pour une tâche.
///
/// Gère la sérialisation / désérialisation depuis Firestore.
/// Conversion vers l'entité domaine via [toEntity].
class TaskModel {
  const TaskModel({
    required this.id,
    required this.title,
    required this.description,
    required this.dueDate,
    required this.status,
    required this.assignedTo,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String title;
  final String description;
  final DateTime dueDate;
  final String status;
  final String assignedTo;
  final String createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory TaskModel.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return TaskModel(
      id: doc.id,
      title: data['title'] as String,
      description: data['description'] as String,
      dueDate: (data['dueDate'] as Timestamp).toDate(),
      status: data['status'] as String,
      assignedTo: data['assignedTo'] as String,
      createdBy: data['createdBy'] as String,
      createdAt: (data['createdAt'] as Timestamp).toDate(),
      updatedAt: (data['updatedAt'] as Timestamp).toDate(),
    );
  }

  Map<String, Object> toJson() => {
        'title': title,
        'description': description,
        'dueDate': Timestamp.fromDate(dueDate),
        'status': status,
        'assignedTo': assignedTo,
        'createdBy': createdBy,
        'createdAt': Timestamp.fromDate(createdAt),
        'updatedAt': Timestamp.fromDate(updatedAt),
      };

  Task toEntity() => Task(
        id: id,
        title: title,
        description: description,
        dueDate: dueDate,
        status: TaskStatus.fromString(status),
        assignedTo: assignedTo,
        createdBy: createdBy,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}

File: apps/angular-app/src/app/features/tasks/models/task.model.ts
export type TaskFirestoreStatus = 'todo' | 'in_progress' | 'done';

export type Task = {
  id: string;
  title: string;
  description: string;
  dueDate: Date;
  status: TaskFirestoreStatus;
  assignedTo: string;
  createdBy: string;
  createdAt: Date;
  updatedAt: Date;
};

export type CreateTaskPayload = {
  title: string;
  description: string;
  dueDate: Date;
  assignedTo: string;
  createdBy: string;
};

export type UpdateTaskPayload = {
  title?: string;
  description?: string;
  dueDate?: Date;
  status?: TaskFirestoreStatus;
  assignedTo?: string;
};

export const TASK_STATUS_OPTIONS: { value: TaskFirestoreStatus; label: string }[] = [
  { value: 'todo', label: 'À faire' },
  { value: 'in_progress', label: 'En cours' },
  { value: 'done', label: 'Terminée' },
];
